src/xrpld/app/tx/detail/Credentials.cpp

```diff
@@ -0,0 +1,470 @@
+    {
+        if (checkNotExpired(sleCred, ctx_.view().info().parentCloseTime))
+        {
+            JLOG(j_.trace()) << "Can't delete non-expired credential.";
+            return tecNO_PERMISSION;
+        }
+    }
+
+    return deleteSLE(view(), sleCred, j_);
+}
+
+// ------- APPLY --------------------------
+
+NotTEC
+CredentialAccept::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureCredentials))
+    {
+        JLOG(ctx.j.trace()) << "featureCredentials is disabled.";
+        return temDISABLED;
+    }
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (!ctx.tx[sfIssuer])
+    {
+        JLOG(ctx.j.trace()) << "Malformed transaction: Issuer field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CredentialAccept::preclaim(PreclaimContext const& ctx)
+{
+    AccountID const subject = ctx.tx[sfAccount];
+    AccountID const issuer = ctx.tx[sfIssuer];
+    auto const credType(ctx.tx[sfCredentialType]);
+
+    if (!ctx.view.exists(keylet::account(issuer)))
+    {
+        JLOG(ctx.j.warn()) << "No issuer: " << to_string(issuer);
+        return tecNO_ISSUER;
+    }
+
+    auto const sleCred =
+        ctx.view.read(keylet::credential(subject, issuer, credType));
+    if (!sleCred)
+    {
+        JLOG(ctx.j.warn()) << "No credential: " << to_string(subject) << ", "
+                           << to_string(issuer) << ", " << credType;
+        return tecNO_ENTRY;
+    }
+
+    if (sleCred->getFieldU32(sfFlags) & lsfAccepted)
+    {
+        JLOG(ctx.j.warn()) << "Credential already accepted: "
+                           << to_string(subject) << ", " << to_string(issuer)
+                           << ", " << credType;
+        return tecDUPLICATE;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialAccept::doApply()
+{
+    AccountID const subject{account_};
+    AccountID const issuer{ctx_.tx[sfIssuer]};
+
+    auto const sleSubject = view().peek(keylet::account(subject));
+    auto const sleIssuer = view().peek(keylet::account(issuer));
+
+    {
+        STAmount const reserve{view().fees().accountReserve(
+            sleSubject->getFieldU32(sfOwnerCount) + 1)};
+        if (mPriorBalance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    auto const credType(ctx_.tx[sfCredentialType]);
+    Keylet const kCred = keylet::credential(subject, issuer, credType);
+    auto const sleCred = ctx_.view().peek(kCred);  // Checked in preclaim()
+
+    if (checkExpired(sleCred, ctx_.view().info().parentCloseTime))
+    {
+        JLOG(j_.trace()) << "Credential is expired: " << sleCred->getText();
+        // delete expired credentials even if the transaction failed
+        CredentialDelete::deleteSLE(ctx_.view(), sleCred, j_);
+        return tecEXPIRED;
+    }
+
+    sleCred->setFieldU32(sfFlags, lsfAccepted);
+    view().update(sleCred);
+
+    adjustOwnerCount(view(), sleIssuer, -1, j_);
```

Comment:
`adjustOwnerCount()` does the `view.update()`. Same few lines below.

Reply:
fixed